app/addons/documents/views.js

```diff
@@ -564,15 +591,79 @@ function(app, FauxtonAPI, Components, Documents, Databases, Views, QueryOptions)
 
       this.toggleTrash();
       this.maybeHighlightAllButton();
+      this.updateExpandButtonLabel();
     },
 
     perPage: function () {
       return this.allDocsNumber.perPage();
+    },
+
+    // a bit dense, but necessary. It decides on what the Expand/Collapse button label should be and takes into
+    // account the default expand/collapse state, what rows are checked and their states
+    updateExpandButtonLabel: function () {
+
+      if (this.defaultDocsStateExpanded) {
+        this.setExpandButtonLabel(false);
+
+        // override the default label and show "Expand" if ALL checked rows have state === collapse (false)
+        if (this.numSelectedOnPage() && this.allCheckedRowsHaveState(false)) {
+          this.setExpandButtonLabel(true);
+        }
+      } else {
+        this.setExpandButtonLabel(true); // default show "Expand"
+
+        // show "Collapse" if ALL checked rows have state === expanded (true)
+        if (this.numSelectedOnPage() && this.allCheckedRowsHaveState(true)) {
+          this.setExpandButtonLabel(false);
+        }
+      }
+    },
```

Comment:
this one is quite complex code - it seems it has a quite high cyclomatic complexity - i would suggest using early returns so the if / else does not get so deeply nested:

```js
    updateExpandButtonLabel: function () {
      this.setExpandButtonLabel(true); // default show "Expand"
      
     // show "Collapse" if ALL checked rows have state === expanded (true)
      if (!this.defaultDocsStateExpanded && this.numSelectedOnPage() && this.allCheckedRowsHaveState(true)) {
        this.setExpandButtonLabel(false);
        return;
      }

      // override the default label and show "Expand" if ALL checked rows have state === collapse (false)
      if (this.numSelectedOnPage() && this.allCheckedRowsHaveState(false)) {
        this.setExpandButtonLabel(true);
      }
      return;

      this.setExpandButtonLabel(false);
    },

```